### PR TITLE
Add max, min, length constraints to StrList parameter 

### LINKED ISSUE #198: Add max, min, length constraints to number list param

### PR DESCRIPTION
**Pull request by krackerman**
fix #198 adding max, min and length constraints to strlist parameter. Wrote tests to get full coverage. updated swagger.js and handler.js

### CODE REVIEW

File: test/resources/swagger.yml [hunk 1]
```diff
@@ -76,6 +90,9 @@ paths:
           enum:
             - enum1
             - enum2
+          minLength: 1
+          maxLength: 2
+          maxItems: 3
```
Comment (simlu) [inline test/resources/swagger.yml]: Do you have a ref to the swagger docs?

Comment (krackerman) [reply test/resources/swagger.yml]: https://swagger.io/docs/specification/data-models/data-types/#string
- minLength
- maxLength
https://swagger.io/docs/specification/data-models/data-types/#array-length
- maxItems

File: src/params/str-list.js [hunk 2]
```diff
@@ -10,6 +10,18 @@ class StrList extends List {
       assert(opts.enums.every(e => typeof e === 'string'));
       this.enums = new Set(opts.enums);
     }
+    if (opts.minListItemLength !== undefined) {
```
Comment (MacMcIrish) [inline src/params/str-list.js]: Doesn't need to be an `if` check.
You're doing the `this.minLIstItemLength !== undefined` in the validation, so it's okay to assign it
```suggestion
  assert(opts.minListItemLength === undefined || Number.isInteger(opts.minListItemLength));
  this.minListItemLength = opts.minListItemLength;
```

The message you're conveying here with the `assert` is that `minListItemLength` can be `undefined` or an `integer`

File: src/params/str-list.js [hunk 3]
```diff
@@ -10,6 +10,18 @@ class StrList extends List {
...
+      assert(typeof opts.minListItemLength === 'number');
```
Comment (MacMcIrish) [inline src/params/str-list.js]: This is the better check for these, we don't want to support list items of length `1.2` ;)
```suggestion
      assert(Number.isInteger(opts.minListItemLength);
```

File: test/handler.js [hunk 4]
```diff
@@ -63,15 +63,21 @@ module.exports.param = api.wrap('POST param', [
   api.IsoDate('isoDateParam', 'json', { required: false }),
   api.UUID('uuidParam', 'json', { required: false }),
   api.List('listParam', 'json', { required: false }),
-  api.StrList('strListParam', 'json', { required: false }),
+  api.StrList('strListParam', 'json', { required: false, enums: ['enum1', 'enum2'] }),
```
Comment (MacMcIrish) [inline test/handler.js]: Why was this updated?

File: test/params/str-list.spec.js [hunk 5]
```diff
@@ -6,6 +6,22 @@ describe('Testing StrList Parameter', () => {
   const jsonParam = api.StrList('list', 'json');
   const queryParamEnum = api.StrList('list', 'query', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
   const bodyParamEnum = api.StrList('list', 'json', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
+  const queryParamInValidConstraints = api.StrList('list', 'query', {
```
Comment (MacMcIrish) [inline test/params/str-list.spec.js]: In Valid reads too close to Invalid, makes it hard to read the tests :)
I thought you were testing invalid constraints below, I'd rename this.

Constraints is also a very general term. If we get our coverage through the unit tests, we're okay.

File: test/params/str-list.spec.js [hunk 6]
```diff
@@ -6,6 +6,22 @@ describe('Testing StrList Parameter', () => {
...
+    minListItemLength: 1,
+    maxListItemLength: 5,
+    maxListLength: 3
+  });
+  const queryParamInValidMin = api.StrList('list', 'query', { minListItemLength: 1 });
+  const bodyParamInValidMin = api.StrList('list', 'json', { minListItemLength: 1 });
+  const queryParamInValidMax = api.StrList('list', 'query', { maxListItemLength: 5 });
+  const bodyParamInValidMax = api.StrList('list', 'json', { maxListItemLength: 5 });
+  const queryParamInValidLength = api.StrList('list', 'query', { maxListLength: 2 });
+  const bodyParamInValidLength = api.StrList('list', 'json', { maxListLength: 2 });
```
Comment (MacMcIrish) [inline test/params/str-list.spec.js]: Wrap each of these in their own describe, it's messy to have to jump up and down from the tests to the definitions to see why / if a test should be passing or failing

File: src/swagger.js [hunk 7]
```diff
@@ -38,6 +38,9 @@ module.exports = (endpoints, existing) => {
           type: 'string',
           enum: [...p.enums]
         },
+        p.minListItemLength === undefined ? {} : { minLength: p.minListItemLength },
+        p.maxListItemLength === undefined ? {} : { maxLength: p.maxListItemLength },
+        p.maxListLength === undefined ? {} : { maxItems: p.maxListLength },
```
Comment (simlu) [inline src/swagger.js]:
```suggestion
        p.maxLength === undefined ? {} : { maxItems: p.maxLength },
```
And then can be removed

File: src/swagger.js [hunk 8]
```diff
@@ -38,6 +38,9 @@ module.exports = (endpoints, existing) => {
           type: 'string',
           enum: [...p.enums]
         },
+        p.minListItemLength === undefined ? {} : { minLength: p.minListItemLength },
+        p.maxListItemLength === undefined ? {} : { maxLength: p.maxListItemLength },
```
Comment (simlu) [inline src/swagger.js]:
```suggestion
        p.maxItemLength === undefined ? {} : { maxLength: p.maxItemLength },
```
same for above

File: src/params/str-list.js [hunk 9]
```diff
@@ -10,6 +10,15 @@ class StrList extends List {
       assert(opts.enums.every(e => typeof e === 'string'));
       this.enums = new Set(opts.enums);
     }
+
+    assert(opts.minListItemLength === undefined || Number.isInteger(opts.minListItemLength));
+    this.minListItemLength = opts.minListItemLength;
```
Comment (simlu) [inline src/params/str-list.js]: rename as discussed below

File: src/params/str-list.js [hunk 10]
```diff
@@ -10,6 +10,15 @@ class StrList extends List {
...
+
+    assert(opts.maxListItemLength === undefined || Number.isInteger(opts.maxListItemLength));
+    this.maxListItemLength = opts.maxListItemLength;
+
+    assert(opts.maxListLength === undefined || Number.isInteger(opts.maxListLength));
+    this.maxListLength = opts.maxListLength;
```
Comment (simlu) [inline src/params/str-list.js]: this should probably be on list?

File: src/params/list.js [hunk 11]
```diff
@@ -12,6 +13,9 @@ class List extends Abstract {
         { type: 'boolean' }
       ]
     };
+
+    assert(opts.maxLength === undefined || Number.isInteger(opts.maxLength));
+    this.maxLength = opts.maxLength;
```
Comment (simlu) [inline src/params/list.js]: while we are at it, lets add a minLength

File: src/swagger.js [hunk 12]
```diff
@@ -38,6 +38,9 @@ module.exports = (endpoints, existing) => {
           type: 'string',
           enum: [...p.enums]
         },
+        p.minItemLength === undefined ? {} : { minLength: p.minItemLength },
+        p.maxItemLength === undefined ? {} : { maxLength: p.maxItemLength },
+        p.maxLength === undefined ? {} : { maxItems: p.maxLength },
```
Comment (simlu) [inline src/swagger.js]:
```suggestion
        p.maxItems === undefined ? {} : { maxItems: p.maxLength },
```
see below

File: src/params/list.js (hunk 11 above)
Comment (simlu) [reply src/params/list.js]: minItems

File: test/params/str-list.spec.js [hunk 13]
```diff
@@ -70,4 +70,88 @@ describe('Testing StrList Parameter', () => {
...
+    it('Testing invalid max-length of list item in json param', () => {
+      expect(() => bodyParamInValidMax.get({
+        body: {
+          list: ['123', '123456']
+        }
+      })).to.throw('Invalid Value for json-Parameter "list" provided.');
+    });
+  });
+
+  describe('Testing optional parameter "maxItems"', () => {
```
Comment (simlu) [inline test/params/str-list.spec.js]: wrong file

**Pull request merged**